### PR TITLE
Disable spurious autolog

### PR DESCRIPTION
**Pull request by BenWilson2**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Disables autologging of sklearn models when transformers Trainer or SetFitTrainer trains. Without disabling this, the internal sklearn model that is created for evaluation purposes causes a new run to be created, the sklearn model logged, and useless parameters and metrics to be logged to this new run. 
This PR disables that logging.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [X] Manual tests (describe details, including test results, below)

Local testing with multiple different training API paradigms.

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Training a transformers model with autologging turned on will no longer create a new run with the internal sklearn model used within the transformers Trainer object when the `train()` method is called.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/transformers.py [hunk 1]
```diff
@@ -1998,3 +1999,53 @@ def _convert_cast_lists_from_np_back_to_list(data):
...
+            return original(*args, **kwargs)
+
+    try:
+        import setfit
+
+        safe_patch(
+            FLAVOR_NAME, setfit.SetFitTrainer, "train", functools.partial(train), manage_run=False
+        )
+    except ImportError:
+        pass
```
Comment (harupy) [inline mlflow/transformers.py:2040]: Can we use `contextlib.suppress`?

Comment (BenWilson2) [reply mlflow/transformers.py]: good idea :) changed

File: mlflow/utils/autologging_utils/__init__.py [hunk 2]
```diff
@@ -475,6 +476,36 @@ def disable_autologging():
...
+        for module in flavors_to_disable
+    }
+    for flavor in flavors_to_disable:
+        autolog_func = getattr(mlflow, flavor)
+        autolog_func.autolog(disable=True)
+    yield None
+    for flavor, state in existing_states.items():
+        if state is not None and not state:
+            autolog_func = getattr(mlflow, flavor)
+            autolog_func.autolog(disable=False)
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply mlflow/utils/autologging_utils/__init__.py]: great idea! Updated

File: mlflow/utils/autologging_utils/__init__.py [hunk 3]
```diff
@@ -475,6 +476,34 @@ def disable_autologging():
...
+                               executing another flavor's autologging to prevent spurious run
+                               logging of unrelated models, metrics, and parameters.
+    """
+    enabled_flavors = []
+    for flavor in flavors_to_disable:
+        if not autologging_is_disabled(flavor):
+            enabled_flavors.append(flavor)
+            autolog_func = getattr(mlflow, flavor)
+            autolog_func.autolog(disable=True)
+    yield None
```
Comment (harupy) [inline mlflow/utils/autologging_utils/__init__.py:501]:
```suggestion
    yield
```

Comment (harupy) [inline mlflow/utils/autologging_utils/__init__.py:501]:
```suggestion
    yield
```

File: tests/transformers/test_transformers_autolog.py [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+    with pytest.raises(MlflowException, match="Run with id"):
+        mlflow.last_active_run()
+    preds = setfit_trainer.model(["Jim, I'm a doctor, not an archaeologist!"])
+    assert len(preds) == 1
+
+
+def test_transformers_autolog_adheres_to_global_behavior_using_trainer(transformers_trainer):
+    mlflow.transformers.autolog()
+
+    transformers_trainer.train()
```
Comment (harupy) [inline tests/transformers/test_transformers_autolog.py:303]: Does this log a run?

File: tests/transformers/test_transformers_autolog.py [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+
+
+def test_transformers_trainer_does_not_autolog_sklearn(transformers_trainer):
+    mlflow.autolog()
+
+    transformers_trainer.train()
+
+    last_run = mlflow.last_active_run()
+    assert last_run.data.metrics is not None
+    assert last_run.data.params is not None
```
Comment (harupy) [inline tests/transformers/test_transformers_autolog.py:279]: If we replace `mlflow.autolog()` with `mlflow.sklearn.autolog()`, would that autolog an sklearn model?

File: tests/transformers/test_transformers_autolog.py (hunk 4 above)
Comment (harupy) [reply tests/transformers/test_transformers_autolog.py:303]: I see, `transformers.Trainer` automatically uses the mlflow callback.

File: tests/transformers/test_transformers_autolog.py [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+    assert len(preds) == 1
+
+
+def test_transformers_autolog_adheres_to_global_behavior_using_trainer(transformers_trainer):
+    mlflow.transformers.autolog()
+
+    transformers_trainer.train()
+
+    last_run = mlflow.last_active_run()
+    assert last_run.data.metrics is not None
```
Comment (harupy) [inline tests/transformers/test_transformers_autolog.py:306]: What does this line check?

File: tests/transformers/test_transformers_autolog.py [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+    with mlflow.start_run(experiment_id=exp_id) as run:
+        model = sklearn.cluster.KMeans()
+        X, y = iris_data
+        model.fit(X, y)
+
+    # Assert that only the run info is logged
+    logged_sklearn_data = mlflow.get_run(run.info.run_id)
+
+    assert not logged_sklearn_data.data.params
+    assert not logged_sklearn_data.data.metrics
```
Comment (harupy) [inline tests/transformers/test_transformers_autolog.py:425]:
```suggestion
    assert logged_sklearn_data.data.params == {}
    assert logged_sklearn_data.data.metrics == {}
```

File: tests/transformers/test_transformers_autolog.py [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+    assert len(preds) == 2
+    assert all(x["score"] > 0 for x in preds)
+
+
+def test_active_autolog_no_setfit_logging_followed_by_successful_sklearn_autolog(
+    iris_data, setfit_trainer
+):
+    mlflow.autolog()
+
+    exp_id = mlflow.create_experiment("setfit_with_sklearn")
```
Comment (harupy) [inline tests/transformers/test_transformers_autolog.py:324]: Curious why we need to create a new experiment?

File: tests/transformers/test_transformers_autolog.py (hunk 5 above)
Comment (BenWilson2) [reply tests/transformers/test_transformers_autolog.py:279]: It won't. To make it clear for this behavior, I changed the explicit override to the global autolog to enable autolog for sklearn in this test

File: tests/transformers/test_transformers_autolog.py (hunk 4 above)
Comment (BenWilson2) [reply tests/transformers/test_transformers_autolog.py:311]: Changed the test validation for logging to make this much more clear with setting new experiments and validating the count of generated runs and their contents.

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**